tasks/src/main/scala/dagr/tasks/ScatterGather.scala

```diff
@@ -0,0 +1,175 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package dagr.tasks
+
+import dagr.core.tasksystem.{Pipeline, Task}
+
+import scala.collection.mutable.ListBuffer
+
+/**
+ * Provides a set of objects and classes for composing Scatter/Gather tasks and pipelines.
+ */
+object ScatterGather {
+
+  /**
+   * A type of Task that, when run, partitions some kind of input, and produces one or more outputs
+   * of type A that should be parallelized (scattered) over.
+   */
+  trait Scatterer[A] extends Task {
+    def scatters: Seq[A]
+  }
+
+  /** Provides a factory method to create a Scatter object from a Scatterer. */
+  object Scatter {
+    def apply[A](scatterer: Scatterer[A]): Scatter[A] = {
+      new ScattererWrapper(scatterer)
+    }
+  }
+
+
+  /**
+   * One of the two main types that implement Scatter/Gather, Scatter provides methods
+   * for mapping the currently scattered type to a new type by running a task in parallel
+   * on each of the scattered parts.
+   */
+  sealed trait Scatter[A] extends Task {
+    /**
+     * Produces a new Scatter from the existing scatter. Takes a function that maps an individual
+     * object of type A to a Task of type B.  During scatter/gather operation this function will
+     * be invoked with each A, to manufacture tasks of type B.
+     *
+     * The resulting Scatter[B] can be further mapped or gathered (or both!).
+     */
+    def map[B <: Task](f: A => B) : Scatter[B]
+
+    /**
+     * Produces a Gather object which will be used at runtime to manufacture a Task of type B
+     * which can gather things of type A.
+     */
+    def gather[B <: Task](f: Seq[A] => B) : Gather[A,B]
+  }
+
+
+  /**
+   * Represents a gather operation in a scatter/gather pipeline.
+   */
+  sealed trait Gather[A, B <: Task] extends Task {
+    private[ScatterGather] var sources: Option[Seq[A]] = None
```

Comment:
That's only partly true.  More specifically the place where `GatherThings` is going to deposit it's result needs to be known ahead of time.  `gatheredThing` could easily be a simple container object into which `GatherThings` places something (e.g. an `AtomicReference[Thing]`).  But I see your point.  Let's log an issue for this and implement it when we need it?  You're right that's it's totally possible.